### ISSUE TITLE
Mixing of .in and .interaction samples

### ISSUE THREAD

**Issue by Matistjati** [Kattis/problem-package-format#263]
Would it be reasonable that a problem being interactive means you should only be allowed to have .interaction samples (no .in)?

If not, should you ever be allowed to mix .in and .interaction? has this *ever* occured?

**Comment by niemela** [comment]
@austrin @jsannemo @simonlindholm @ghamerly Thoughts?

**Comment by niemela** [comment]
> If not, should you ever be allowed to mix .in and .interaction? has this _ever_ occured?

To clarify, the question is whether there can be a problem where some samples have `.interaction` and other do not and only have `.in`. Not if a single sample can have a `.interaction` *and* a `.in`, which is clearly true.

This is in fact specified in the "Interactive Problems" section. But it took way longer to find and figure this out. The relevant sections could clearly take some clarifications.

(Or, we might be too tired?)

**Comment by RagnarGrootKoerkamp** [comment]
Some more comments. (See also #265)

## What is shown in the statement

For each testcase (defined by the presence of either a `.in` or `.interaction` file)
1. If there is a `.interaction`, show that.
2. If there is a `.in` + `.out`, show those.
3. If there is a `.in` + `.ans`, show those.

We could restrict this to always require the same type, i.e., never allow mixing `.interaction` with `.in+.ans/.out`, and we could even disallow mixing 2. and 3. for consistency (although I'm not sure how useful that is).

Do we allow having some `.interaction` _with_ corresponding `.in`, and some _without_ corresponding `.in`?

Note: there are technically interactive problems that we present as usual in-out, so requiring a `.interaction` for interactive problems is too strong. On the other hand, non-interactive problems should never have a `.interaction` file.

Having both a `.out` and `.interaction` file for a single testcase is forbidden; both are only for presentation purposes, and only one can be shown.

A `.out` file is allowed for both default validation and custom validation. For default validation, it can be used to use a different number of significant digits, while for custom validation it is an example of an actual output, whereas `.ans` is only input for the output validator.

## What is available to contestants as download

default validation: give .in and .out (or .ans if not present)
custom validation: give .in and .out (or .ans if not present)
interactive:
 - if no `.interaction` => give `.in` and `.out`
 - if there is only `.interaction` => give nothing, but put things in attachments/
 - if there is `.interaction` and `.in/ans` => give nothing, but put things in attachments/

**Comment by simonlindholm** [comment]
> Fake interactive problems / generated input problems

Do we have examples of this actually being used in practice? There's never a case where you actually strictly need this, right? I can say I've never felt the need, and it feels like it just risks providing worse UX (e.g. having judges show confusing UI and not being able to show failing test case input/output in the same way as for other non-interactive problems) and adding a lot of bug potential around stdin/stdout fd closures, EOF checking and other termination behavior.

> To clarify, the question is whether there can be a problem where some samples have .interaction and other do not and only have .in.

FWIW, https://github.com/zehnsechs/egoi-2024-testdata/tree/main/day1/gardendecorations/data/sample had this, because we wanted to have a sample test case that would be run by Kattis while also splitting the .interaction file into three because it was a multi-run problem. I'm not sure how we're envisioning samples to work for multi-run problems.

**Comment by niemela** [comment]
> I'm not sure how we're envisioning samples to work for multi-run problems.

multi-pass will use `.interaction` files (https://www.kattis.com/problem-package-format/spec/2023-07-draft.html#multi-pass-validation).

(Was that the answer you were looking for?)

**Comment by simonlindholm** [comment]
Ah, thanks, that makes sense.

**Comment by RagnarGrootKoerkamp** [comment]
> Do we have examples of this actually being used in practice?

Yes, we had multiple such 'generated input' problems for BAPC, in particular where we guarantee that the input is random, and hence regenerated on each re-submission.

**Comment by RagnarGrootKoerkamp** [comment]
Reopening, since there are still some unresolved discussions in #291.

I think one thing that also isn't really specified is whether for custom output validation and interactive problems, we require that for each test case in `data/sample/statement`, the **same** set of files is present, rather than an independent set of valid files.

**Comment by simonlindholm** [comment]
> Yes, we had multiple such 'generated input' problems for BAPC, in particular where we guarantee that the input is random, and hence regenerated on each re-submission.

Interesting. Is the problem package available for any of them? I do feel like that kind of setup is inadvisable, and it's better to keep the test data static while still giving a guarantee that it was generated at random.

**Comment by RagnarGrootKoerkamp** [comment]
See problem `L` here: https://2022.bapc.eu/bapc/problems.pdf
(You can download the sources via the homepage.)

**Comment by niemela** [comment]
@RagnarGrootKoerkamp Could you summarize what's still remaining in this issue? (Maybe even close this and create a new one, depending on what would be cleaner?)

**Comment by RagnarGrootKoerkamp** [comment]
From my reading of the #291 discussions:
- for interactive problems, we default to an empty `data/sample/download`, and any files for download must be explicitly listed there. (Usually, `data/sample/statement`, or else just `data/sample` is what is provided for `.in` and `.ans` download.)

- Consider to allow `...` lines for `statement/*.interaction`. But one can already do `>...` and `<...` which actually get shown on the right side of the interaction, so probably not very useful.

- Clarify that "for each test case, `data/sample/statement` must contain ...` .

**Comment by RagnarGrootKoerkamp** [comment]
> I think one thing that also isn't really specified is whether for custom output validation and interactive problems, we require that for each test case in `data/sample/statement`, the **same** set of files is present, rather than an independent set of valid files.

Not sure about this. Requiring the same set of files guarantees more consistency for teams and a bit more work to verify correctness, but probably that simplifies the logic around things, so I guess I'm inclined to require this for now and we can always loosen later if needed.

**Comment by RagnarGrootKoerkamp** [comment]
> I don't quite understand this point? I think you are saying that this is what we should do, but it's not what the spec says yet?

yes indeed; that what it _should_ say.

> I assume that "allow ..." means that you can do a ... that is shown between messages and not on either side. I think that would make sense.

SGTM

> What is unclear?

See here: https://github.com/Kattis/problem-package-format/pull/291#discussion_r1722229013

> same set of files

as in: if you have two samples `1` and `2`,  `data/sample/statement/1.EXT` must exist if and only if `data/sample/statement/2.EXT` exists, and likewise for more extensions.

> We should probably list .out and .interaction files in the top of the "Test data" section (and note that they are only used in data/sample.

SGTM

> They should either both be allowed in data/sample or neither.

Yeah, even before coming to that final sentence of that paragraph I was already thinking: it doesn't really make sense to have `data/sample/*.interaction` now, does it? Cleaner to require those to be in `data/sample/statement/*.interaction` indeed.

**Comment by niemela** [comment]
I think #375 should close this. AFAICT, these things are not covered:

> See here: [#291 (comment)](https://github.com/Kattis/problem-package-format/pull/291#discussion_r1722229013)

I don't we want to say that no other files may exist. E.g. I would want to allow `README`s pretty much anywhere in the package.

> * Consider to allow `...` lines for `statement/*.interaction`

I think we should at least consider this. That said, they are equally useful in a non-interaction flow. I.e. if you have a `.ans.statement` that is an abbreviated `.ans`. How could we support `...` in `.ans.statement`? I don't think we want to do that? Could we allow the use of `.interaction` (where `...` would be special handled) in a non-interactive non-multi-pass problem? If we do, should it be called something else than `.interaction`?

**Comment by mzuenni** [comment]
I think in `.ans.statement` it is weird, `...` is a line which could just occur. In interaction it would be fine but for me just `>...` or `<...` is also fine.

**Comment by niemela** [comment]
> I think in `.ans.statement` it is weird, `...` is a line which could just occur.

Agreed. That would make an input of `...` on a line invalid. Although, only for sample *and* if you need to override the sample output, but still... weird.

> In interaction it would be fine but for me just `>...` or `<...` is also fine.

So, what should I do if I have an input or output that is "too long" and I want an abbreviated `.in.statement` or `.ans.statement`? We could allow `.interaction` files for *all* types of problems and then I could have a `.interaction` as such:
```
<(beginning of input)
...
<(end of input
>(beginning of output)
...
>(end of input)
```
Then a system could (should?) know to show `.interaction` files where all `<` lines are before all `>` lines as normal input/output blocks, because that's really what it is.

?

**Comment by niemela** [comment]
> on the other hand this sounds like a can of worms to me. We could add special handling for a lot of things but i don't think its a good idea to start with this...

I can't think of anything we should do *berfore* this. Can you?

I mean, I can totally see "we shouldn't do this", but I don't see the "there is some other special handling we should add before this". Am I reading too much between the lines?

> In this scenario just writing `...` and saying in text that some part is skipped is sufficient?

Sure, that's what we'll do if we don't support something better.

We already support `>`, `<`, and `---`, adding `...` to this set doesn't feel to terrible (or particularly can of worm-y).

**Comment by mzuenni** [comment]
> I can't think of anything we should do berfore this. Can you?

it was meant to read "we should not start with more special handlings"

**Comment by niemela** [comment]
> it was meant to read "we should not start with more special handlings"

Ok, sorry, i *did* read to much between the lines then.
